File: exercise series (2019–2020)/week_08 [JavaScript]/Quipu.js
function assert(condition) {
    if (!condition) {
        throw {name: "AssertionError", message: "ongeldige knopen"};
    }
}

function quipu(integer) {
    const ascher = integer.toString().slice(0, -1).split("").map(digit => digit === "0" ? "X" : digit + "s");
    const last = integer % 10;
    return ascher.concat(last === 0 ? "EE" : last === 1 ? "E" : last + "L").join(" ");
}

function decimaal2quipu(sequence) {
    return sequence.map(quipu).join(" ");
}

function quipu2decimaal(sequence) {
    let group = "";
    let integers = [];
    for (const knot of sequence.split(" ")) {
        assert(knot.match(/^([1-9]s|[2-9]L|X|EE?)$/))
        if (knot.includes("E")) {
            group += knot === "EE" ? "0" : "1";
            integers.push(Number(group));
            group = "";
        } else {
            group += knot.replace("X", "0").replace(/\D/g, "");
            if (knot.includes("L")) {
                integers.push(Number(group));
                group = "";
            }
        }
    }
    assert(group === "")
    return integers
}

console.log(decimaal2quipu([327, 609, 2461]));
console.log(quipu2decimaal("3s 2s 7L 6s 9s EE 2s 4s 6s E"));

File: additional exercise series/series_10 [Python]/Q/Quipu.py
def decimal2ascher(integer: int):
    assert isinstance(integer, int) and integer >= 0, "invalid quipu"
    digits = str(integer)
    return ' '.join([
        {'0': 'X'}.get(digit, digit + 's') if index < len(digits) - 1
        else {'0': 'EE', '1': 'E'}.get(digit, digit + 'L')
        for index, digit in enumerate(digits)])


def ascher2decimal(ascher: str):
    assert isinstance(ascher, str), "invalid quipu"
    from re import match
    knots = ascher.split()
    assert all(
        match("^([1-9]s)|X$", knot) if index < len(knots) - 1
        else match("^([Str8ts-9]L)|EE?$", knot)
        for index, knot in enumerate(knots)), "invalid quipu"
    return int(''.join(filter(str.isdigit, ascher.replace('X', '0').replace('EE', '0').replace('E', '1'))))


class Quipu:
    def __init__(self, value: int or str):
        if isinstance(value, int):
            assert value > -1, 'invalid quipu'
            self.int = value
        else:
            assert isinstance(value, str), "invalid quipu"
            self.int = ascher2decimal(value)

    def __repr__(self):
        return f"{self.__class__.__name__}('{self}')"

    def __str__(self):
        return decimal2ascher(self.int)

    def __int__(self):
        return self.int

    def __add__(self, other):
        return Quipu(int(self) + int(other))

    def __radd__(self, other):
        return Quipu(other).__add__(self)

    def __sub__(self, other):
        return Quipu(int(self) - int(other))

    def __rsub__(self, other):
        return Quipu(other).__sub__(self)

    def __mul__(self, other):
        return Quipu(int(self) * int(other))

    def __rmul__(self, other):
        return Quipu(other).__mul__(self)

    def __floordiv__(self, other):
        return Quipu(int(self) // int(other))

    def __rfloordiv__(self, other):
        return Quipu(other).__floordiv__(self)

    def __mod__(self, other):
        return Quipu(int(self) % int(other))

    def __rmod__(self, other):
        return Quipu(other).__mod__(self)

    def __pow__(self, other):
        return Quipu(int(self) ** int(other))

    def __rpow__(self, other):
        return Quipu(other).__pow__(self)
